pkg/containers/slice_queue.go

```diff
@@ -0,0 +1,109 @@
+package containers
+
+import "sync"
+
+// SliceQueue is a FIFO queue implemented
+// by a Go slice.
+type SliceQueue[T any] struct {
+	mu    sync.Mutex
+	elems []T
+
+	C chan struct{}
+
+	pool *sync.Pool
+}
+
+// NewSliceQueue creates a new SliceQueue.
+func NewSliceQueue[T any]() *SliceQueue[T] {
+	return &SliceQueue[T]{
+		C:    make(chan struct{}, 1),
+		pool: &sync.Pool{},
+	}
+}
+
+func (q *SliceQueue[T]) Add(elem T) {
+	q.mu.Lock()
+
+	signal := false
+	if len(q.elems) == 0 {
+		signal = true
+		if q.elems == nil {
+			q.elems = q.allocateSlice()
+			q.elems = q.elems[:0]
+		}
+	}
+
+	q.elems = append(q.elems, elem)
+	q.mu.Unlock()
+
+	if signal {
+		select {
+		case q.C <- struct{}{}:
+		default:
+		}
+	}
+}
+
+func (q *SliceQueue[T]) Pop() (T, bool) {
+	q.mu.Lock()
+
+	var zero T
+	if len(q.elems) == 0 {
+		q.mu.Unlock()
+		return zero, false
+	}
+
+	ret := q.elems[0]
+	q.elems[0] = zero
+	q.elems = q.elems[1:]
+
+	if len(q.elems) == 0 {
+		q.freeSlice(q.elems)
+		q.elems = nil
+	} else {
```

Comment:
Does it mean that if we don't consume the queue, we will block  in the  queue.Pop even the queue is not empty?
Is it by design? 
It seems that we only want a notifier mechanism to  tell us that queue is not empty.